packages/augur-core/src/contracts/para/AMMExchange.sol

```diff
@@ -8,236 +8,251 @@ import 'ROOT/ICash.sol';
+        liquidityConstant = calculateLiquidityConstant();
+        // CONSIDER: convert min(poolInvalid, poolYes, poolNo) to DAI by selling complete sets. Selling complete sets incurs Augur fees, maybe we should let the user sell the sets themselves if they want to pay the fee?
+    }
+
+    // Spend cash for shares.
+    // You will receive as many shares as you can afford for the given cash.
+    // If that isn't as many shares as _minSharesReceived then this reverts.
+    function enterPosition(uint256 _amountInCash, bool _buyYes, uint256 _minSharesReceived) public returns (uint256) {
+        uint256 _setsToBuy = _amountInCash.div(numTicks);
+        uint256 _position = rateEnterPosition(_setsToBuy, _buyYes);
+
+        assert(_position >= _minSharesReceived, "Not enough cash to buy at least the minimum requested shares.");
+
+        // materialize the final result of the simulation
+        cash.transferFrom(msg.sender, address(this), _setsToBuy.mul(numTicks));
+        if (_buyYes) {
+            shareTransfer(address(this), msg.sender, _setsToBuy, 0, _position);
+        } else {
+            shareTransfer(address(this), msg.sender, _setsToBuy, _position, 0);
+        }
+
+        return _position;
+    }
+
+    // How many shares you get if you buy complete sets then sell the other side.
+    function rateEnterPosition(uint256 _setsToBuy, bool _buyYes) public view returns (uint256) {
+        (uint256 _poolInvalid, uint256 _poolNo, uint256 _poolYes) = shareBalances(address(this));
+
+        // simulate the user buying complete sets directly from the exchange
+        _poolInvalid = _poolInvalid.subS(_setsToBuy, "AugurCP: The pool doesn't have enough INVALID tokens to fulfill the request.");
+        _poolNo = _poolNo.subS(_setsToBuy, "AugurCP: The pool doesn't have enough NO tokens to fulfill the request.");
+        _poolYes = _poolYes.subS(_setsToBuy, "AugurCP: The pool doesn't have enough YES tokens to fulfill the request.");
+
+        require(_poolInvalid > 0, "AugurCP: The pool doesn't have enough INVALID tokens to fulfill the request.");
+        require(_poolNo > 0, "AugurCP: The pool doesn't have enough NO tokens to fulfill the request.");
+        require(_poolYes > 0, "AugurCP: The pool doesn't have enough YES tokens to fulfill the request.");
+
+        // simulate user swapping YES to NO or NO to YES
+        uint256 _poolConstant = poolConstant(_poolYes, _poolNo);
+        if (_buyYes) {
+            // yesToUser + poolYes - poolConstant / (poolNo + _setsToBuy)
+            return _setsToBuy.add(_poolYes.sub(_poolConstant.div(_poolNo.add(_setsToBuy))));
+        } else {
+            return _setsToBuy.add(_poolNo.sub(_poolConstant.div(_poolYes.add(_setsToBuy))));
+        }
+    }
+
+    // If you do not have complete sets then you must have more shares than _setsToSell because you will be swapping
+    // some of them to build complete sets.
+    function exitPosition(uint256 _setsToSell) external {
```

Comment:
This should have a `minCashReceived` parameter to protect users from frontrunners.

Reply:
You'll always receive the same amount of cash for a given amount of complete sets. However, the risk here is in how many shares you swap away to build complete sets. I'll add `_maxSharesSwappedAway`.

Looks like we're going to switch to doing the quadratic equation on-chain though so this is probably moot.